staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -0,0 +1,253 @@
+func (s *openAPIAggregator) updateOpenAPISpec() error {
+	if s.openAPIService == nil {
+		return nil
+	}
+	specToServe, err := s.buildOpenAPISpec()
+	if err != nil {
+		return err
+	}
+	return s.openAPIService.UpdateSpec(specToServe)
+}
+
+// inMemoryResponseWriter is a http.Writer that keep the response in memory.
+type inMemoryResponseWriter struct {
+	header   http.Header
+	respCode int
+	data     []byte
+}
+
+func newInMemoryResponseWriter() *inMemoryResponseWriter {
+	return &inMemoryResponseWriter{header: http.Header{}}
+}
+
+func (r *inMemoryResponseWriter) Header() http.Header {
+	return r.header
+}
+
+func (r *inMemoryResponseWriter) WriteHeader(code int) {
+	r.respCode = code
+}
+
+func (r *inMemoryResponseWriter) Write(in []byte) (int, error) {
+	r.data = append(r.data, in...)
+	return len(in), nil
+}
+
+func (r *inMemoryResponseWriter) String() string {
+	s := fmt.Sprintf("ResponseCode: %d", r.respCode)
+	if r.data != nil {
+		s += fmt.Sprintf(", Body: %s", string(r.data))
+	}
+	if r.header != nil {
+		s += fmt.Sprintf(", Header: %s", r.header)
+	}
+	return s
+}
+
+func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if ctx, ok := s.contextMapper.Get(req); ok {
+			s.contextMapper.Update(req, request.WithUser(ctx, info))
+		}
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// Load openAPI spec from /swagger.json of given handler.
+func (s *openAPIAggregator) loadOpenAPISpec(handler http.Handler) (*spec.Swagger, error) {
+	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
+	handler = request.WithRequestContext(handler, s.contextMapper)
+
+	req, err := http.NewRequest("GET", "/swagger.json", nil)
+	if err != nil {
+		return nil, err
+	}
+	writer := newInMemoryResponseWriter()
+	handler.ServeHTTP(writer, req)
+
+	switch writer.respCode {
+	case http.StatusOK:
+		openApiSpec := &spec.Swagger{}
+		if err := json.Unmarshal(writer.data, openApiSpec); err != nil {
+			return nil, err
+		}
+		return openApiSpec, nil
+	default:
+		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
+	}
+}
+
+func max(i, j int32) int32 {
+	if i > j {
+		return i
+	}
+	return j
+}
+
+// loadApiServiceSpec loads OpenAPI spec for the given API Service and then updates aggregator's spec.
+func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) error {
+
+	// Ignore local services
+	if apiService.Spec.Service == nil {
+		return nil
+	}
+
+	openApiSpec, err := s.loadOpenAPISpec(handler)
+	if err != nil {
+		return err
+	}
+	aggregator.FilterSpecByPaths(openApiSpec, []string{"/apis/" + apiService.Spec.Group + "/"})
+
+	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
+		serviceName: apiService.Name,
+		spec:        openApiSpec,
+		priority:    max(apiService.Spec.VersionPriority, apiService.Spec.GroupPriorityMinimum),
```

Comment:
My understanding of these two values was completely wrong. I've changed sorting algorithm to fix this.